src/lib.rs

```diff
@@ -0,0 +1,419 @@
+        // Check for a '=' delimiter and use the second half of the split as the format,
+        // otherwise remove next arg which is now at the same position as the original flag.
+        let format = if let Some((_, format)) = arg.split_once('=') {
+            format.to_string()
+        } else {
+            input.cargo_opts.remove(pos)
+        };
+
+        // Non-json formats are not supported so the executable exits.
+        if !format.starts_with("json") {
+            eprintln!("error: non-JSON `message-format` is not supported");
+            process::exit(1);
+        } else {
+            format
+        }
+    } else {
+        // Default to 'json-render-diagnostics'
+        DEFAULT_MESSAGE_FORMAT.to_string()
+    }
+}
+
+/// Build the elf that will be used to create other 3ds files.
+/// The command built from [`make_cargo_build_command`] is executed
+/// and the messages from the spawned process are parsed and returned.
+pub fn build_elf(
+    cmd: CargoCommand,
+    message_format: &str,
+    args: &Vec<String>,
+) -> (ExitStatus, Vec<Message>) {
+    let mut command = make_cargo_build_command(cmd, message_format, args);
+    let mut process = command.spawn().unwrap();
+    let command_stdout = process.stdout.take().unwrap();
+
+    let mut tee_reader;
+    let mut stdout_reader;
+    let buf_reader: &mut dyn BufRead = if message_format == DEFAULT_MESSAGE_FORMAT {
+        stdout_reader = BufReader::new(command_stdout);
+        &mut stdout_reader
+    } else {
+        // The user presumably cares about the message format, so we should
+        // copy stuff to stdout like they expect. We can still extract the executable
+        // information out of it that we need for 3dsxtool etc.
+        tee_reader = BufReader::new(TeeReader::new(command_stdout, io::stdout()));
+        &mut tee_reader
+    };
+
+    let messages = Message::parse_stream(buf_reader)
+        .collect::<io::Result<_>>()
+        .unwrap();
+
+    (process.wait().unwrap(), messages)
+}
+
+/// Create the cargo build command, but don't execute it.
+/// If there is no pre-built std detected in the sysroot, `build-std` is used.
+pub fn make_cargo_build_command(
+    cmd: CargoCommand,
+    message_format: &str,
+    args: &Vec<String>,
+) -> Command {
+    let rust_flags = env::var("RUSTFLAGS").unwrap_or_default()
+        + &format!(
+            " -L{}/libctru/lib -lctru",
+            env::var("DEVKITPRO").expect("DEVKITPRO is not defined as an environment variable")
+        );
+    let cargo = env::var("CARGO").unwrap_or_else(|_| "cargo".to_string());
+    let sysroot = find_sysroot();
+    let mut command = Command::new(cargo);
+
+    if !sysroot.join("lib/rustlib/armv6k-nintendo-3ds").exists() {
+        eprintln!("No pre-build std found, using build-std");
+        command.arg("-Z").arg("build-std");
+    }
+
+    command
+        .env("RUSTFLAGS", rust_flags)
+        .arg(&cmd.to_string())
```

Comment:
This regressed. When the user says `cargo 3ds run` we don't actually want to execute `cargo run`. Instead, we execute `cargo build` and run it via `3dslink`.

Instead of calling `to_string`, there should probably be a method to get the cargo command to run.

Reply:
Since they are passed in as a Ref<OsString>, it needed to be some sort of string. My implementation for this is the `Display` trait on `CargoCommand`. Simply changing what is written to the formatter fixes this. If there is an idea of an alternate method I'd gladly look into it.

Reply:
It's very weird to use `Display` for this, especially since the string doesn't always match the enum (`Run` returns `build`).

I suggested making a method on `CargoCommand` which would return the command which should be used. Or, inline it as part of `make_cargo_build_command`.

Reply:
Should be all set with this now. Let me know if there are any other changes pertaining to this!